Introduce new DSA switch framework

yangbolu1991 opened the pull request:
This PR is to introduce new DSA switch framework.  the RFC for new DSA architecuture is at discussion: https://github.com/zephyrproject-rtos/zephyr/discussions/87091

This PR is first update according to new DSA arthiteture in RFC, the purpose and changes are as below.
Also a basic doc is added in patch
https://github.com/zephyrproject-rtos/zephyr/pull/87045/commits/45bf6388a339449bd97b7706c8aaf5ac5aeb3485


    - Aligned to Linux DSA framework which has been already mature framework
      for many years, For now in zephyr, the DSA compoenents were splited
      as: switch, port, master(not need driver file for now), slave, and tag.
      Seperated drivers were used for maintaining and developing new
      features.

    - The unified dts bindings (aligned to linux) were supported. The port
      driver would parse DTS to decide the port type (master port, slave
      port, or cpu port) to set up the switch. All the ports registered as
      standard ethernet devices. (dsa port and dsa switch tree was not
      supported.)

    - How to add DSA device driver based on the framework? Following dts bindings to describe
      the switch device, all the device driver needing to do is providing dsa_api implementaion and private
      data, and calling below initialization.

        DSA_INIT_INSTANCE(n, _dapi, data)

    - For switch tag case, recv/xmit helpers in dsa_api could be used for
      taging/untagging. No modified ethernet drivers.
      For no-tag type case, ethernet driver of master port should support
      packet injection/extraction for slave ports leaving NULL recv/xmit.

    The dsa_nxp_imx_netc.c driver will be the first example of the new DSA
    framework.

    The future work for DSA will be supporting bridge for ports. We may align
    Linux to give users two options to use DSA device:

    - Standalone mode: the switch ports work as regular ethernet ports.

    - Bridge mode: switch mode with virtual bridge device which could be
      assigned IP address.

yangbolu1991 committed:
dts: bindings: dsa: add common bindings file

Added common bindings file for dsa, and adapted NETC to it.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: eth_nxp_imx_netc: support no-tag type DSA conduit

Added support for no-tag type DSA conduit.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
doc: connectivity: networking: add DSA document

Added DSA document.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

pdgendt commented on drivers/ethernet/Kconfig.dsa:
```diff
@@ -4,20 +4,26 @@
 
 if NET_DSA
 
+config NET_DSA_LEGACY
+	bool "Distributed Switch Architecture support for legacy device"
+	depends on ETH_DSA_SUPPORT_LEGACY
```
Probably want to select the `DEPRECATED` option.

```suggestion
config NET_DSA_LEGACY
	bool "Distributed Switch Architecture support for legacy device"
	select DEPRECATED
	depends on ETH_DSA_SUPPORT_LEGACY
```

pdgendt commented on doc/connectivity/networking/dsa.rst:
```diff
@@ -0,0 +1,125 @@
+Host interface
+**************
+
+Host interface network devices use regular and unmodified ethernet driver working
+as DSA master port, which manages the switch through processor.
```
Similar to linux we should phase out the term "master" and prefer "conduit", similar instead of "slave" use "user".

From https://docs.kernel.org/networking/dsa/dsa.html

> NB: for the past 15 years, the DSA subsystem had been making use of the terms “master” (rather than “conduit”) and “slave” (rather than “user”). These terms have been removed from the DSA codebase and phased out of the uAPI.

yangbolu1991 replied:
Right. Will convert to new terms.
Thanks.

lmajewski reviewed (commented):
@yangbolu1991 - Thank you for your work put to port relevant pieces of Linux DSA driver to Zephyr.

Generally, for me the code looks OK. It is also great that you added the detailed documentation for DSA subsystem to Zephyr.

Currently the https://docs.zephyrproject.org/latest/boards/segger/ip_k66f/doc/index.html board (including its sample LLDP protocol program)  uses the `NET_DSA_LEGACY` but I think that it shall finally support the new framework.

JiafeiPan commented on doc/connectivity/networking/dsa.rst:
```diff
@@ -0,0 +1,125 @@
+           |               Switch              |
+           |+------+ +------+ +------+ +------+|
+           || swp0 | | swp1 | | swp2 | | swp3 ||
+           ++------+-+------+-+------+-+------++
```
I think we could the rx/tx diagram in https://github.com/zephyrproject-rtos/zephyr/discussions/87091, but not use this diagram which is from Linux kernel document.

yangbolu1991 replied:
Okay, will replace it.
Thanks.

JiafeiPan commented on subsys/net/l2/ethernet/dsa/dsa.c:
```diff
@@ -1,84 +1,24 @@
-
 /*
- * Copyright (c) 2018 Intel Corporation
+ * Copyright 2025 NXP
```
According to our discussion, let's add a "dsa_" prefix in order to follow the same naming rules in Zephyr driver, for example, use "dsa_core.c dsa_port.c dsa_salve.c ..."

Another suggestion, in order to make the old architectrue code and new architecuture code works simulteneously in new architecuture developing period,  let's don't modify the old file directly, we could create new architeccture code in each new clean file, so that we could define which files can be compiled in cmake according to CONFIG_NET_DSA_DEPRECATED.

yangbolu1991 replied:
It makes sense. That will make code much clear. Will do that.
Thank you very much.

yangbolu1991 committed:
net: dsa: make current DSA core driver obsolete

Current DSA TX/RX way was hard-coded in ethernet devices driver
with ETH_DSA_SUPPORT and NET_DSA.
This patch is to make such way obsolete, as we actually will support
a better DSA framework to handle this in NET/DSA core driver.
To make legacy devices not affected, below options are used instead.

- ETH_DSA_SUPPORT_DEPRECATED
- NET_DSA_DEPRECATED

Once the legacy devices are converted to new DSA framework, this code
could be removed.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
include: ethernet: rename dsa master/slave to conduit/user

Renamed dsa master/slave to conduit/user for capabilities.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: rename master/slave to conduit/user

Renamed master/slave to conduit/user.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
samples: net: dsa: convert to new dsa core driver

Converted to new dsa core driver. And adapted mimxrt1180_evk to it.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 committed:
drivers: ethernet: dsa_nxp_imx_netc: adapt to new DSA framework

Adapted to use new DSA framework. And related platforms converted
in samples too.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
> @yangbolu1991 - Thank you for your work put to port relevant pieces of Linux DSA driver to Zephyr.
> 
> Generally, for me the code looks OK. It is also great that you added the detailed documentation for DSA subsystem to Zephyr.
> 
> Currently the https://docs.zephyrproject.org/latest/boards/segger/ip_k66f/doc/index.html board (including its sample LLDP protocol program) uses the `NET_DSA_LEGACY` but I think that it shall finally support the new framework.

Thank you so much for your support :)

yangbolu1991 commented:
Updated to v2. Changes include,
- Selected DEPRECATED for obsolete options
- Converted to new terms dsa conduit/user
- Updated doc DSA TX/RX process picture.
- Added new dsa_xxx drivers keeping the original dsa.c/dsa.h unmodified. Built control with CONFIG_NET_DSA_DEPRECATED option for the old/new dsa core driver.

Thanks.

maass-hamburg commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+	const struct dsa_port_config dsa_##dsa##_port_##port##_config = {                          \
+		.mac_addr = DT_PROP_OR(port, local_mac_address, {0}),                              \
+		.port_idx = DT_REG_ADDR(port),                                                     \
+		.phy_dev = (COND_CODE_1(DT_NODE_HAS_PROP(port, phy_handle),                        \
+					DEVICE_DT_GET(DT_PHANDLE(port, phy_handle)),               \
+					NULL)),                                                    \
```
```suggestion
		.phy_dev = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(port, phy_handle)),               \
```

yangbolu1991 replied:
Will convert to use it, thanks.

maass-hamburg commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+					DEVICE_DT_GET(DT_PHANDLE(port, phy_handle)),               \
+					NULL)),                                                    \
+		.phy_mode = DT_PROP_OR(port, phy_connection_type, ""),                             \
+		.ethernet_connection = (COND_CODE_1(DT_NODE_HAS_PROP(port, ethernet),              \
```
similar here

maass-hamburg commented on dts/bindings/dsa/dsa.yaml:
```diff
@@ -0,0 +1,26 @@
+    - name: pinctrl-device.yaml
+    - name: ethernet-controller.yaml
+      property-allowlist:
+        - local-mac-address
```
why no ``zephyr,random-mac-address``

yangbolu1991 replied:
Let me add the random mac support.
Thanks.

maass-hamburg commented on include/zephyr/net/ethernet.h:
```diff
@@ -953,12 +961,23 @@ enum ethernet_hw_caps net_eth_get_hw_capabilities(struct net_if *iface)
 	const struct device *dev = net_if_get_device(iface);
 	const struct ethernet_api *api = (struct ethernet_api *)dev->api;
+	enum ethernet_hw_caps caps = 0;
+#if defined(CONFIG_NET_DSA)
+#if !defined(CONFIG_NET_DSA_DEPRECATED)
```
```suggestion
#if defined(CONFIG_NET_DSA) && !defined(CONFIG_NET_DSA_DEPRECATED)
```

yangbolu1991 replied:
will fix, thanks.

jukkar commented on drivers/ethernet/Kconfig.nxp_enet:
```diff
@@ -12,7 +12,7 @@ config ETH_NXP_ENET
 	select ARM_MPU if CPU_CORTEX_M7
```
Typo in commit msg of commit `net: dsa: make current DSA core driver obsolete`
s/deivices/devices/
Perhaps the sentence could say:
`Once the legacy devices are converted to new DSA framework, this code could be removed`

yangbolu1991 replied:
Thanks, will fix.

jukkar commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+/*
```
In commit `net: dsa: introduce new DSA framework` some typos in commit message
s/compoenents/components/
s/implementaion/implementation/

yangbolu1991 replied:
Thanks, will fix.

jukkar commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+	/** DSA specific API callbacks - filled in the switch IC driver */
+	struct dsa_api *dapi;
+
+	/** Number of ports in the DSA switch */
+	uint8_t num_ports;
+
+	/** Number of initialized ports in the DSA switch */
+	uint8_t init_ports;
+
+	/** Instance specific data */
+	void *prv_data;
```
Minor nit, but could we change the ordering of these fields to avoid hole in the struct because of alignment. This would not affect the size of the struct atm, but could if new fields are added to it. So suggesting
```
	/** Instance specific data */
	void *prv_data;

	/** Number of ports in the DSA switch */
	uint8_t num_ports;

	/** Number of initialized ports in the DSA switch */
	uint8_t init_ports;
```

jukkar commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+	int (*switch_setup)(const struct dsa_context *dsa_ctx);
+};
+
+enum dsa_port_type {
+	NON_DSA_PORT,
+	DSA_CONDUIT_PORT,
+	DSA_USER_PORT,
+	DSA_CPU_PORT,
+	DSA_PORT,
```
Doxygen documentation missing for these.

yangbolu1991 replied:
I should have moved them out of INTERNAL_HIDDEN.
Will do that.
Thanks.

jukkar replied:
If these are not to be documented, then the `/** @cond INTERNAL_HIDDEN */` line below should be moved before this enum, and then the previous hidden doc block can be combined with this one.
But if you meant that these are to be documented, then the documentation is still missing.
I did not quite understand your comment here.

I think the `/** @cond INTERNAL_HIDDEN */` blocks are incorrectly set now. You have documentation of the APIs but they are marked now as hidden which prevents documentation generation. Please readjust these checks.

yangbolu1991 replied:
Sorry, I just realized it's my code style problem.
Going through the header file, the only thing needs to public is one API `dsa_user_get_iface` used in sample.
Others were used only in drivers. Will fix issues.
Thank you very much.

jukkar commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,272 @@
+	DSA_PORT,
+};
+
+struct dsa_port_config {
+	uint8_t mac_addr[6];
+	const int port_idx;
+	const struct device *phy_dev;
+	const char *phy_mode;
+	const struct device *ethernet_connection;
+	const struct pinctrl_dev_config *pincfg;
+};
```
Same here. If this is considered "private" struct, then you need to add
```
/** @cond INTERNAL_HIDDEN */
...
/** @endcond */
```
guards that will not generate documentation for it.

yangbolu1991 replied:
Thanks. I made it in INTERNAL_HIDDEN as I thought it's used only by driver.

jukkar commented on include/zephyr/net/ethernet.h:
```diff
@@ -26,8 +26,12 @@
 #include <zephyr/net/ethernet_vlan.h>
 #include <zephyr/net/ptp_time.h>
 
+#if defined(CONFIG_NET_DSA)
```
Usually it is not needed to add ifdef for including public header files. This helps to avoid clutter *.c code with ifdefs.

yangbolu1991 replied:
Get it. Let me remove it.

yangbolu1991 replied:
Although I removed ifdef for CONFIG_NET_DSA here, the twister issues showed that dsa_core.h including ``zephyr/drivers/pinctrl.h`` also must require ``pinctrl_soc.h``.
I actually intended to make pinctrl as optional in dsa, should I wrapper pinctrl header/code with ifdef?
Or there may be better idea?

Thank you very much.

jukkar commented on include/zephyr/net/ethernet.h:
```diff
@@ -132,7 +136,7 @@ struct net_eth_addr {
  * store tail tag.
  */
-#if defined(CONFIG_NET_DSA_DEPRECATED)
+#if defined(CONFIG_NET_DSA)
```
So we could avoid this ifdef if in dsa_core.h we could define `DSA_TAG_SIZE` to 0 if DSA is not enabled. This would make the code more readable.

yangbolu1991 replied:
Right. Actually both dsa.h and dsa_core.h already defined ``DSA_TAG_SIZE`` by checking CONFIG_DSA_TAG_SIZE.
So, let me remove the ifdef directly.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa_port.c:
```diff
@@ -0,0 +1,142 @@
+		eth_ctx_conduit->dsa_port = DSA_CONDUIT_PORT;
+	}
+
+	if (strcmp(cfg->phy_mode, "internal") == 0) {
```
Sometimes external security checker tools might start to complain about these "unsafe" string functions. So in order to avoid that in the future, we could be proactive here and change this to
```
if (strncmp(cfg->phy_mode, "internal", sizeof("internal") - 1) == 0) {
```

yangbolu1991 replied:
Thanks, I will convert to use that.

jukkar commented on doc/connectivity/networking/index.rst:
```diff
@@ -18,3 +18,4 @@ operation of the stacks and how they were implemented.
    network_tracing.rst
    api/index.rst
    conn_mgr/index.rst
+   dsa.rst
```
The image is nice and it is ok to have it in png format. But if we ever needs to adjust the image in the future, could you also provide the original file in svg format?

yangbolu1991 replied:
Sure. Will provide svg.

jukkar reviewed (commented):
Overall looks pretty good. I have could of minor comments.

yangbolu1991 committed:
boards: nxp: mimxrt1180_evk: convert to random mac addresses for switch

Converted to use random mac addresses for switch.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented on subsys/net/ip/net_core.c:
```diff
@@ -499,6 +499,12 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	int ret;
 
+	/* DSA driver handles first to untag and to redirect to slave interface. */
+#if defined(CONFIG_NET_DSA)
+#if !defined(CONFIG_NET_DSA_DEPRECATED)
```
okay, will change it, thanks.

yangbolu1991 commented on subsys/net/ip/packet_socket.c:
```diff
@@ -25,14 +30,20 @@ enum net_verdict net_packet_socket_input(struct net_pkt *pkt, uint8_t proto)
+	if (dsa_port_is_conduit(net_pkt_iface(pkt))) {
+		return NET_CONTINUE;
+	}
+#endif
 #endif
```
okay, will add, thanks.

yangbolu1991 commented on include/zephyr/net/ethernet.h:
```diff
@@ -192,10 +196,10 @@ enum ethernet_hw_caps {
 	/** DSA switch slave port */
-	ETHERNET_DSA_SLAVE_PORT		= BIT(15),
+	ETHERNET_DSA_USER_PORT		= BIT(15),
 
 	/** DSA switch master port */
-	ETHERNET_DSA_MASTER_PORT	= BIT(16),
+	ETHERNET_DSA_CONDUIT_PORT	= BIT(16),
```
okay, will add, thanks.

yangbolu1991 commented on drivers/ethernet/dsa/Kconfig:
```diff
@@ -4,20 +4,27 @@
-config ETH_DSA_SUPPORT
+config ETH_DSA_SUPPORT_DEPRECATED
 	bool
+	select DEPRECATED
```
Get it. Will remove, thanks.

yangbolu1991 commented:
Updated to v3. Changes include,
- Improved code and code style per suggestion of  @jukkar  and @maass-hamburg 
- Supported ``zephyr,random-mac-address`` in dsa driver and converted RT1180 NETC DSA to use it.
- Used svg image in doc and fixed ``INTERNAL_HIDDEN`` in ``dsa_core.h``.
- Fixed no tag type DSA conduit support in NETC ETH driver. (should be wrapperred by CONFIG_NET_DSA)

Thanks a lot.

jukkar commented:
@yangbolu1991 I have the mimxrt1180 board and tried this PR. Couple of questions:
1. I am seeing 7 Ethernet network interfaces created. I am seeing 5 Ethernet sockets in the board. I suppose the 6th interface is the conduit interface, but where that 7th interface coming from?
2. I am guessing that the conduit network interface is not having any direct network traffic. We should mark it so in order to enforce this in higher level.
3. Could we mark the conduit port to be down if no other port is connected, or is this a chicken-and-egg problem so does it need to be up before other ports notice that the cable is connected?

While testing this, I have some changes to the DSA sample and I could tweak the conduit interface too (bullet 2 above). Are you ok if I push my changes to your branch?

jukkar committed:
samples: net: dsa: Enable DHCPv4 and start it when interface is up

Enable DHCPv4 when network interface goes up. This makes it easier
to test IP connectivity with DSA.

Signed-off-by: Jukka Rissanen <jukka.rissanen@nordicsemi.no>
Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

maass-hamburg commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,274 @@
+
+/** @cond INTERNAL_HIDDEN */
+
+#define DSA_PORT_MAX_COUNT 8
```
This should be a Kconfig or even better derived from the DT

yangbolu1991 replied:
Right, I will change to derive from DT.
Thanks.

yangbolu1991 commented:
> @yangbolu1991 I have the mimxrt1180 board and tried this PR. Couple of questions:
> 
>     1. I am seeing 7 Ethernet network interfaces created, but there are 5 Ethernet sockets in the board. I suppose the 6th interface is the conduit interface, but where is that 7th interface coming from?
> 
>     2. I am guessing that the conduit network interface is not having any direct network traffic. We should mark it so in order to enforce this in higher level.
> 
>     3. Could we mark the conduit port to be down if no other port is connected, or is this a chicken-and-egg problem so does it need to be up before other ports notice that the cable is connected?
> 
> 
> While testing this, I have some changes to the DSA sample and I could tweak the conduit interface too (bullet 2 above). Are you ok if I push my changes to your branch?

1. On RT1180, there are 5 switch ports (including 1 internal port), and 2 enetc ports (including 1 internal port). The two internal ports are connected inside SoC. So, you see 7 ports in zephyr, and 5 external port on board.
All the ports and connection were cleanly described in dts. See,
https://github.com/zephyrproject-rtos/zephyr/blob/main/dts/arm/nxp/nxp_rt118x.dtsi#L627
https://github.com/zephyrproject-rtos/zephyr/blob/main/boards/nxp/mimxrt1180_evk/mimxrt1180_evk.dtsi#L79

if you try ``net iface``, the switch ports should have name swp0 to swp4. The capability showed dsa user and dsa conduit on ports correspondingly.

In Linux, the CPU port will be hidden.

2. For the conduit port, it should be up first before any user ports are up, as it manages frame injection/extraction on all user ports. It's a special port and user couldn't use it as normal ethernert port.

3. Yes, conduit port should be up first, and then the user ports. Actually I think the up should be MAC up, but not phy link up.
That's actually what linux is doing. Below is an linux example to use DSA switch, eno2 is the conduit port.

```
ifconfig eno2 up
ip link add name switch type bridge
ip link set switch up
ip link set swp0 master switch && ip link set swp0 up
ip link set swp1 master switch && ip link set swp1 up
ip link set swp2 master switch && ip link set swp2 up
ip link set swp3 master switch && ip link set swp3 up
```

Just feel free to push your patches also to the branch.
Thank you very much.

jukkar commented:
> Just feel free to push your patches also to the branch.

I was not able to do that, the branch does not allow it. Anyway, my changes can be found here
https://github.com/jukkar/zephyr/tree/people/nxp/dsa-support

jukkar commented:
> * the switch ports should have name swp0 to swp4.

Hmm, I think the network interfaces were called eth0-eth7 (cannot verify atm). Anyway, the names can be set by the driver as it sees fit using `net_if_set_name(struct net_if *iface, const char *buf);` in the interface init API function.

yangbolu1991 commented:
> > * the switch ports should have name swp0 to swp4.
> 
> Hmm, I think the network interfaces were called eth0-eth7 (cannot verify atm). Anyway, the names can be set by the driver as it sees fit using `net_if_set_name(struct net_if *iface, const char *buf);` in the interface init API function.

Sorry, I mean device name. It seems net iface show it only when link up.
Let me try to set the interface name, that will be clear.
Thanks.

yangbolu1991 commented:
> > Just feel free to push your patches also to the branch.
> 
> I was not able to do that, the branch does not allow it. Anyway, my changes can be found here https://github.com/jukkar/zephyr/tree/people/nxp/dsa-support

I went through the two patches.

For the below patch, actually I think it may be early to add.
In the future, when the bridge layer is supported for DSA. The IP should be assigned to the virtual bridge for switch ports.
As a switch, the ports should not have IP. When ports are not bridged, they may work as standalone mode as normal ethernet interfaces. (we can set ip.)
But for now, I tested it, and it seemed worked fine. The switch function was fine, and the switch port basic ping with the IP worked fine. So, I added the patch.

```
e3960771daf samples: net: dsa: Enable DHCPv4 and start it when interface is up
```

For the second patch, I haven't added it. I am thinking if we can mark in proper place in core driver.

```
f2fa4153980 drivers: ethernet: dsa_nxp_imx_netc: Disable IPv4/6 for conduit port

```

Thanks.

jukkar commented:
> For the second patch, I haven't added it. I am thinking if we can mark in proper place in core driver.
> 
> ```
> f2fa4153980 drivers: ethernet: dsa_nxp_imx_netc: Disable IPv4/6 for conduit port
> ```

I would have thought that this is more important from these two. This patch will prevent IP address assignment to the conduit interface and also prevent upper stack from trying to send data via that interface. Without this patch, the `net iface` command will show the interface as a "normal" network interface that is up but does not have proper IP addresses etc, except that it will have multicast addresses added to it.

jukkar commented:
Let me explain what I mean
Without my `drivers: ethernet: dsa_nxp_imx_netc: Disable IPv4/6 for conduit port` patch
```
uart:~$ net iface 7
Hostname: dsa
Default interface: 1


Interface eth1 (0x1400c9f8) (Ethernet) [7]
===================================
Link addr : 00:00:00:01:03:00
MTU       : 1500
Flags     : AUTO_START,IPv4,IPv6
Device    : ethernet@60b40000 (0x38036620)
Status    : oper=UP, admin=UP, carrier=ON
Ethernet capabilities supported:
        RX checksum offload
        10 Mbits
        100 Mbits
        1 Gbits
        MAC address filtering
        DSA conduit port
Ethernet PHY device: <none> (0)
IPv6 unicast addresses (max 5):
        fe80::200:ff:fe01:300 autoconf preferred infinite
IPv6 multicast addresses (max 5):
        ff02::1
        ff02::1:ff01:300
IPv6 prefixes (max 2):
        <none>
IPv6 hop limit           : 64
IPv6 base reachable time : 30000
IPv6 reachable time      : 26117
IPv6 retransmit timer    : 0
IPv4 unicast addresses (max 1):
        <none>
IPv4 multicast addresses (max 2):
        224.0.0.1
IPv4 gateway : 0.0.0.0
DHCPv4 lease time : 0
DHCPv4 renew time : 0
DHCPv4 server     : 0.0.0.0
DHCPv4 requested  : 0.0.0.0
DHCPv4 state      : disabled
DHCPv4 attempts   : 0
```
The conduit interface has all the IP features enabled even if it cannot handle any IP traffic itself.

With the patch applied we have
```
uart:~$ net iface 7
Hostname: dsa
Default interface: 1


Interface eth1 (0x1400c9f8) (Ethernet) [7]
===================================
Link addr : 00:00:00:01:03:00
MTU       : 1500
Flags     : AUTO_START
Device    : ethernet@60b40000 (0x38036658)
Status    : oper=UP, admin=UP, carrier=ON
Ethernet capabilities supported:
        RX checksum offload
        10 Mbits
        100 Mbits
        1 Gbits
        MAC address filtering
        DSA conduit port
Ethernet PHY device: <none> (0)
IPv6 not enabled for this interface.
IPv4 not enabled for this interface.
```
And IP networking is no longer possible and the upper layers will not utilize this network interface for anything IP related.

jukkar commented on subsys/net/ip/net_if.c:
```diff
@@ -428,7 +431,12 @@ static inline void init_iface(struct net_if *iface)
 	net_if_flag_set(iface, NET_IF_IPV6);
 #endif
-
+#if defined(CONFIG_NET_DSA)
```
We have been trying to limit adding driver specific configuration to generic code. So in this respect, this is not the proper place. We could have init function in `subsys/net/l2/ethernet/dsa/dsa.c` and which could be called by `ethernet_init()` in `subsys/net/l2/ethernet/ethernet.c` . That `dsa_init()` could set these things if needed.

yangbolu1991 replied:
okay, thanks for suggestion.
Let me try to do that:)

yangbolu1991 commented on subsys/net/ip/net_core.c:
```diff
@@ -560,6 +560,10 @@ int net_recv_data(struct net_if *iface, struct net_pkt *pkt)
 	int ret;
 
+	/* DSA driver handles first to untag and to redirect to user interface. */
+#if defined(CONFIG_NET_DSA) && !defined(CONFIG_NET_DSA_DEPRECATED)
```
No. All packets which switch could received is through conduit port (the host ethernet interface.)
So, dsa_recv calling dsa_tag_recv will check if the interface is conduit port by if dsa_context exists.
If not, just return.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,299 @@
+#endif
+
+/** DSA context data */
+struct dsa_context {
```
yes. switch is proper.
legacy dsa code used dsa_context. So, let me add one more dsa_switch_context which will be exact.
Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,299 @@
+	 * dsa_context.
+	 */
+	/** Program (set) mac table entry in the DSA switch */
+	int (*switch_set_mac_table_entry)(const struct device *dev, const uint8_t *mac,
```
How about dropping them for now? These are actually from legacy dsa code.
And this PR doesn't give any platform support example for it.
We can consider more when we support these callback in the future.
Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,299 @@
+/**
+ * @brief DSA transmit function
+ *
+ * @param dev Device
```
```Port device to use to transmit```
Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,299 @@
+	const char *phy_mode;
+	const struct device *ethernet_connection;
+#if defined(CONFIG_PINCTRL)
+	const struct pinctrl_dev_config *pincfg;
```
But pin config is indeed not net area. I'm thinking if move to device driver, how we implement in code changes.
Any suggestion?

```
#define DSA_PORT_INIT_INSTANCE(port, dsa)                                                          \
        COND_CODE_1(DT_NUM_PINCTRL_STATES(port),                                                   \
                                (PINCTRL_DT_DEFINE(port);), (EMPTY))                               \
        struct dsa_port_config dsa_##dsa##_port_##port##_config = {                                \
                .use_random_mac_addr = DT_NODE_HAS_PROP(port, zephyr_random_mac_address),          \
                .mac_addr = DT_PROP_OR(port, local_mac_address, {0}),                              \
                .port_idx = DT_REG_ADDR(port),                                                     \
                .phy_dev = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(port, phy_handle)),                    \
                .phy_mode = DT_PROP_OR(port, phy_connection_type, ""),                             \
                .ethernet_connection = DEVICE_DT_GET_OR_NULL(DT_PHANDLE(port, ethernet)),          \
                .pincfg = COND_CODE_1(DT_NUM_PINCTRL_STATES(port),                                 \
                                (PINCTRL_DT_DEV_CONFIG_GET(port)), NULL),                          \
        };                                                                                         \
        NET_DEVICE_INIT_INSTANCE(                                                                  \
                CONCAT(dsa_, dsa, _port_, port),                                                   \
                DEVICE_DT_NAME(port), DT_REG_ADDR(port), dsa_port_initialize, NULL,                \
                                &dsa_context_##dsa, &dsa_##dsa##_port_##port##_config,             \
                                CONFIG_ETH_INIT_PRIORITY, &dsa_eth_api, ETHERNET_L2,               \
                                NET_L2_GET_CTX_TYPE(ETHERNET_L2), NET_ETH_MTU);

#define DSA_INIT_INSTANCE(n, _dapi, data)                                                          \
        struct dsa_context dsa_context_##n = {                                                     \
                .dapi = _dapi,                                                                     \
                .prv_data = (void *)data,                                                          \
                .init_ports = 0,                                                                   \
                .num_ports = DT_INST_CHILD_NUM_STATUS_OKAY(n),                                     \
        };                                                                                         \
        DT_INST_FOREACH_CHILD_STATUS_OKAY_VARGS(n, DSA_PORT_INIT_INSTANCE, n);

```

Thanks.

yangbolu1991 committed:
net: dsa: introduce new DSA framework

The new DSA framework purpose and changes are as below.

- Aligned to Linux DSA framework which has been already mature framework
  for many years, For now in zephyr, the DSA components were splited
  as: switch, port, master(not need driver file for now), slave, and tag.
  Seperated drivers were used for maintaining and developing new
  features.

- The unified dts bindings (aligned to linux) were supported. The port
  driver would parse DTS to decide the port type (master port, slave
  port, or cpu port) to set up the switch. All the ports registered as
  standard ethernet devices. (dsa port and dsa switch tree was not
  supported.)

- How to add DSA device driver based on the framework? All the device
  driver needing to do is providing dsa_spi implementation and private
  data, and calling below initialization.

    DSA_INIT_INSTANCE(n, _dapi, data)

- For switch tag case, recv/xmit helpers in dsa_api could be used for
  taging/untagging. No modified ethernet drivers.
  For no-tag type case, ethernet driver of master port should support
  packet injection/extraction for slave ports leaving NULL recv/xmit.

The dsa_nxp_imx_netc.c driver will be the first example of the new DSA
framework.

The future work for DSA will be supporting bridge for ports. We may align
Linux to give users two options to use DSA device:

- Standalone mode: the switch ports work as regular ethernet ports.

- Bridge mode: switch mode with virtual bridge device which could be
  assigned IP address.

Signed-off-by: Yangbo Lu <yangbo.lu@nxp.com>

yangbolu1991 commented:
Updated to v5. Changes include,
- used dsa_switch_context instead of dsa_context.
- moved dsa port config to device driver to give the configuration.
- moved pinctrl setting to device driver port init.
- Removed switch_set_mac_table_entry/switch_set_mac_table_entry/dsa_register_recv_callback APIs in new dsa driver which copied from legacy code, as these (and the usage in dsa sample) were not following real DSA desgin. This should be via bridge layer in the future.
- Supported disabling ipv4/v6 for conduit port. 
- Other improvement in code.

Thanks a lot.

mmahadevan108 commented on drivers/ethernet/nxp_imx_netc/dsa_nxp_imx_netc.c:
Any reason this is not in the dsa folder

yangbolu1991 replied:
Thanks a lot.
Just checked. The dsa folder was created last month. It missed to move existed dsa netc driver.
To make it perfect, I could move it in another PR once this PR gets accepted, as this PR purpose is to add dsa framework.

```
commit 814421162fcdcc8c7eb5560e606a6028cab4edba
Author: Fin Maaß <f.maass@vogl-electronic.com>
Date:   Mon Mar 17 17:00:53 2025 +0100

    drivers: ethernet: dsa: move to own subdirectory

    move dsa drivers to its own subdirectory

    Signed-off-by: Fin Maaß <f.maass@vogl-electronic.com>
```

Thanks a lot.

yangbolu1991 commented:
Updated to v6. Changes include,
- Rebased to address conflict.
- Dropped hal patch for conduit port support and handled in netc eth driver instead.

Thanks a lot.

yangbolu1991 commented:
Just rebased to address doc conflict.
Thanks.

jukkar commented:
@maass-hamburg please re-review

jukkar commented on boards/nxp/mimxrt1180_evk/mimxrt1180_evk.dtsi:
```diff
@@ -95,7 +95,7 @@
 
 &switch_port0 {
-	local-mac-address = [00 00 00 01 04 00];
+	zephyr,random-mac-address;
```
This board has Ethernet sockets that have factory MAC address marked to them. Why can't we use the factory set MAC addresses here?

yangbolu1991 replied:
I may need to have some investigation on how to do that in both zephyr and platform.
Not sure if any fuse related driver needed?
Could you share any factory set MAC address case support in zephyr?
I'd like to support that seperately for NETC and RT1180 platform in another PR if it's okay.
Thank you very much.

jukkar replied:
For example NXP frdm_rw612 board has both Wi-Fi and Ethernet. The Wi-Fi is properly using factory set MAC address, but the Ethernet is not. I created an issue for it #88336

I have STM nucleo_f767zi board which uses device id by default a mac address. https://github.com/zephyrproject-rtos/zephyr/blob/1e1f94405f138e5e0164dea0bcac400517ab88d5/drivers/ethernet/eth_stm32_hal.c#L846

jukkar replied:
Having a separate PR for setting MAC address to fixed value is ok. I was just wondering why NXP is using random MAC address for several boards I have been using.

yangbolu1991 replied:
I have had some investigation but I have no idea on the reason:(
It looks like the way getting the factory mac address varied from soc/board.
I think for i.MXRT it's from fuse which may read via `hal/nxp/mcux/mcux-sdk-ng/drivers/ocotp` driver.
For some i.MX8 and i.MX9, it's from fuse too, but it's more complicated to read. As the reading must be via talking with EdgeLock Enclave. I can see below drivers in u-boot for that.
```
drivers/misc/imx_ele/fuse.c
drivers/misc/imx_ele/ele_api.c
```
Not sure about other series SoC.
Anyway, I will take some time to make this clear, and try to support in zephyr.
Thanks.

maass-hamburg commented on drivers/ethernet/nxp_imx_netc/dsa_nxp_imx_netc.c:
```diff
@@ -4,281 +4,148 @@
-port_init:
 	/* miiSpeed and miiDuplex will get set correctly when link is up */
-	swt_config->ports[cfg->port_idx].ethMac.miiMode = cfg->phy_mode;
+	swt_config->ports[cfg->port_idx].ethMac.miiMode =
+		(strcmp(cfg->phy_mode, "mii") == 0
+			 ? kNETC_MiiMode
+			 : (strcmp(cfg->phy_mode, "rmii") == 0
+				    ? kNETC_RmiiMode
+				    : (strcmp(cfg->phy_mode, "rgmii") == 0
+					       ? kNETC_RgmiiMode
+					       : (strcmp(cfg->phy_mode, "gmii") == 0
+							  ? kNETC_GmiiMode
+							  : kNETC_RmiiMode))));
```
use with `DT_INST_ENUM_HAS_VALUE()` in the `DSA_NETC_PORT_INST_INIT()`, so we don't have to compare strings at runtime.

yangbolu1991 replied:
@maass-hamburg Thanks.
That's true, but using `DT_INST_ENUM_HAS_VALUE()` returns maybe a device specific value to use, like `kNETC_MiiMode` here.

Then it seems meaningless of the `const char *phy_mode;` definition in header to parsing node string property.
```
struct dsa_port_config {
        uint8_t mac_addr[6];
        const bool use_random_mac_addr;
        const int port_idx;
        const struct device *phy_dev;
        const char *phy_mode;
        const struct device *ethernet_connection;
        /** Instance specific config */
        void *prv_config;
};
```

As phy mode initialization is just once at beginning, I'm not sure if it affects a lot.
Or, should we remove `const char *phy_mode;` in header to let device driver to define?
Thanks.

maass-hamburg replied:
``const char *phy_mode`` would then have to use the same type as ``miiMode``.

yangbolu1991 replied:
Thanks. Let me use `netc_hw_mii_mode_t phy_mode` in device driver to decide during build.

jukkar commented on drivers/ethernet/nxp_imx_netc/eth_nxp_imx_netc.c:
```diff
@@ -26,6 +26,12 @@ const struct device *netc_dev_list[NETC_DRV_MAX_INST_SUPPORT];
 {
 	struct netc_eth_data *data = dev->data;
+	struct net_if *iface_dst = data->iface;
+#if !(defined(FSL_FEATURE_NETC_HAS_SWITCH_TAG) && FSL_FEATURE_NETC_HAS_SWITCH_TAG) && \
+	defined(CONFIG_NET_DSA)
```
I am seeing this ifdef used in many places in this driver. It is somewhat hard to read, so I would suggest making a shorter symbol that can be used instead. So something like this
```
#if !(defined(FSL_FEATURE_NETC_HAS_SWITCH_TAG) && FSL_FEATURE_NETC_HAS_SWITCH_TAG) && \
	defined(CONFIG_NET_DSA)
#define USE_DSA
#elif defined(USE_DSA)
#undef USE_DSA
#endif
```
then the checks would become much simpler and easier to read
```
#if defined(USE_DSA)
....
#endif
```

yangbolu1991 replied:
Great. I will do that.
Thanks.

jukkar commented on samples/net/dsa/src/main.c:
```diff
@@ -24,32 +24,69 @@ static void dsa_iface_find_cb(struct net_if *iface, void *user_data)
+	ARG_UNUSED(iface);
+	ARG_UNUSED(cb);
+
+	if ((mgmt_event & EVENT_MASK) != mgmt_event) {
+		return;
+	}
```
This check is not really needed as you are checking the event few lines below.

yangbolu1991 replied:
Will remove, thanks.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa_port.c:
```diff
@@ -0,0 +1,123 @@
+	const struct device *dev = net_if_get_device(iface);
+	struct dsa_port_config *cfg = (struct dsa_port_config *)dev->config;
+	struct dsa_switch_context *dsa_switch_ctx = dev->data;
+	char name[10];
```
There is a config option for the length, so we can use it here
```
char name[CONFIG_NET_INTERFACE_NAME_LEN];
```

yangbolu1991 replied:
Will do that, thanks.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa_port.c:
```diff
@@ -0,0 +1,123 @@
+	char name[10];
+
+	/* Set interface name */
+	sprintf(name, "swp%d", cfg->port_idx);
```
And here we should never use the unsafe sprintf function. Please change this to
```
snprintkf(name, sizeof(name), ....);
```

yangbolu1991 replied:
Will convert to snprintf.
Thanks.

jukkar commented on subsys/net/l2/ethernet/dsa/dsa_user.c:
```diff
@@ -0,0 +1,30 @@
+	struct ethernet_context *eth_ctx = net_if_l2_data(iface);
+	struct dsa_switch_context *dsa_switch_ctx;
+
+	if (!eth_ctx || !eth_ctx->dsa_switch_ctx) {
```
We try to avoid this style of checks and have been converting these to explicit NULL checks in other parts of the code. So better to do here
```
if (eth_ctx == NULL || eth_ctx->dsa_switch_ctx == NULL) {
```
Please check other such invocations in this PR and adjust the code accordingly.

yangbolu1991 replied:
Thanks, will fix all.

yangbolu1991 commented:
Updated to v7. Changes include,
- Got phy mode during build instead of run-time in dsa netc driver.
- Fixed API doc in `dsa_core.h`.
-  Other code improvement per @jukkar  suggestion.

Thank you very much.

kartben commented on doc/connectivity/networking/dsa.rst:
```diff
@@ -0,0 +1,127 @@
+- ``DSA_SWITCH_INST_INIT(n, _dapi, data, fn)``
+- ``DSA_PORT_INST_INIT(port, n, cfg)``
+
+.. code-block:: console
```
```suggestion
.. code-block:: c
```

yangbolu1991 replied:
Will fix, thanks!

kartben commented on doc/connectivity/networking/dsa.rst:
```diff
@@ -0,0 +1,127 @@
+This is copied from Linux DSA documentation. It applies to zephyr too. Although conduit port and
+cpu port exposed as ethernet device in zephyr, they are not able to be used.
+
+.. code-block:: console
```
??

maybe you meant `.. note::`

yangbolu1991 replied:
Will fix, thanks!

kartben reviewed (changes requested):
some comments related to docs.
Also, did was this presented/discussed in the Architecture WG?

dleach02 commented:
> some comments related to docs. Also, did was this presented/discussed in the Architecture WG?

It was discussed in the Networking WG meeting on 4/14. The meeting was recorded.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
```diff
@@ -0,0 +1,191 @@
+/**
+ * @brief Distributed Switch Architecture (DSA)
+ * @defgroup dsa_core Distributed Switch Architecture (DSA)
+ * @since 2.5
```
I once hesitated to use 2.5 or not previously, as the legacy dsa.h existed.
Let me use 4.2 I think it makes sense since it's new file added.
Thanks.

yangbolu1991 commented on doc/connectivity/networking/dsa.rst:
```diff
@@ -0,0 +1,127 @@
+For device driver, all needed to prepare are ``dsa_api``, private data if has, and ``dsa_port_config``.
+The macro functions could be utilized. And below is an example of i.MX NETC.
+
+- ``DSA_SWITCH_INST_INIT(n, _dapi, data, fn)``
+- ``DSA_PORT_INST_INIT(port, n, cfg)``
```
Okay, will do that.
I put it in hidden doxygen block was because I thought it's used only by drivers. And the doc was describing driver usage.
Let me move them and related structures out. I think maybe it's better for developer to understand and for cross reference.
Thanks.

yangbolu1991 commented on include/zephyr/net/dsa_core.h:
Right... I had intended to hidden most.
But let me expose APIs/structure related to device driver support. That may be useful for device driver developer to read doc.
Thanks.
